components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -764,62 +959,80 @@ private void sendResponse(HttpServletRequest req, HttpServletResponse resp, Stri
                 }
                 out.print(soapResponse);
             } else {
+                generateSamlPostPageFromFile(resp, acUrl, response, relayState, authenticatedIdPs,
+                        SAMLSSOConstants.SAML_RESP);
+            }
 
-            String finalPage = null;
-            String htmlPage = IdentitySAMLSSOServiceComponent.getSsoRedirectHtml();
-            String pageWithAcs = htmlPage.replace("$acUrl", acUrl);
-            String pageWithAcsResponse = pageWithAcs.replace("<!--$params-->", "<!--$params-->\n" + "<input type='hidden' name='SAMLResponse' value='" + Encode.forHtmlAttribute(response) + "'/>");
-            String pageWithAcsResponseRelay = pageWithAcsResponse;
+        } else {
+            generateSamlPostPageFromTemplate(resp, acUrl, response, relayState, authenticatedIdPs,
+                    SAMLSSOConstants.SAML_RESP);
+        }
+    }
 
-            if(relayState != null) {
-                pageWithAcsResponseRelay = pageWithAcsResponse.replace("<!--$params-->", "<!--$params-->\n" + "<input type='hidden' name='RelayState' value='" + Encode.forHtmlAttribute(relayState)+ "'/>");
-            }
+    private void generateSamlPostPageFromTemplate(HttpServletResponse resp, String acUrl, String samlMessage,
+                                                  String relayState, String authenticatedIdPs, String samlMessageType)
+            throws IOException {
 
-            if (authenticatedIdPs == null || authenticatedIdPs.isEmpty()) {
-                finalPage = pageWithAcsResponseRelay;
-            } else {
-                finalPage = pageWithAcsResponseRelay.replace(
-                        "<!--$additionalParams-->",
-                        "<input type='hidden' name='AuthenticatedIdPs' value='"
-                                + Encode.forHtmlAttribute(authenticatedIdPs) + "'/>");
-            }
+        PrintWriter out = resp.getWriter();
```

Comment:
Without writing each and every line to response, it's better to build the whole response and write to output in a single go.

Reply:
Fixed with 3bbc008dd576ca0169fde3b49fe4f505a99200ea